src/main/scala/org/broadinstitute/dsde/firecloud/service/TSVFileSupport.scala

```diff
@@ -0,0 +1,113 @@
+package org.broadinstitute.dsde.firecloud.service
+
+import akka.actor.Actor
+import org.broadinstitute.dsde.firecloud.model.AttributeUpdateOperations.{AddUpdateAttribute, AttributeUpdateOperation, RemoveAttribute}
+import org.broadinstitute.dsde.firecloud.model.ModelJsonProtocol.AttributeFormat
+import org.broadinstitute.dsde.firecloud.model._
+import org.broadinstitute.dsde.firecloud.service.PerRequest.PerRequestMessage
+import org.broadinstitute.dsde.firecloud.utils.{TSVLoadFile, TSVParser}
+import spray.http.StatusCodes
+import spray.http.StatusCodes._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+import spray.json._
+/**
+  * Created by ansingh on 11/14/16.
+  */
+trait TSVFileSupport extends Actor {
+
+  private implicit val impPlainAttributeFormat: AttributeFormat = new AttributeFormat with PlainArrayAttributeListSerializer
+
+  /**
+    * Attempts to parse a string into a TSVLoadFile.
+    * Bails with a 400 Bad Request if the TSV is invalid. */
+  def withTSVFile(tsvString:String)(op: (TSVLoadFile => Future[PerRequestMessage]))(implicit ec: ExecutionContext): Future[PerRequestMessage] = {
+    Try(TSVParser.parse(tsvString)) match {
+      case Failure(regret) => Future(RequestCompleteWithErrorReport(BadRequest, regret.getMessage))
+      case Success(tsvFile) => op(tsvFile)
+    }
+  }
+
+  def checkNumberOfRows(tsv: TSVLoadFile, rows: Int)(op: => Future[PerRequestMessage])(implicit ec: ExecutionContext): Future[PerRequestMessage] = {
+    if ((tsv.tsvData.length + (if (tsv.headers.isEmpty) 0 else 1)) != rows) {
+      Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest,
+        "Your file does not have the correct number of rows. There should be " + rows.toString))
+    } else {
+      op
+    }
+  }
+
+  def checkRowDistinct( tsv: TSVLoadFile )(op: => Future[PerRequestMessage])(implicit ec: ExecutionContext): Future[PerRequestMessage] = {
+    val attributeNames = Seq(tsv.headers.head.stripPrefix("workspace:")) ++ tsv.headers.tail
+    if (attributeNames.size != attributeNames.distinct.size) {
+      Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Duplicated attribute keys are not allowed"))
+    } else {
+      op
+    }
+  }
+
+  /**
+    * Bail with a 400 Bad Request if the first column of the tsv has duplicate values.
+    * Otherwise, carry on. */
+  def checkFirstColumnsDistinct( tsv: TSVLoadFile )(op: => Future[PerRequestMessage])(implicit ec: ExecutionContext): Future[PerRequestMessage] = {
+    val entitiesToUpdate = tsv.tsvData.map(_.headOption.get)
+    val distinctEntities = entitiesToUpdate.distinct
+    if ( entitiesToUpdate.size != distinctEntities.size ) {
+      Future( RequestCompleteWithErrorReport(BadRequest,
+        "Duplicated entities are not allowed in TSV: " + entitiesToUpdate.diff(distinctEntities).distinct.mkString(", ")) )
+    } else {
+      op
+    }
+  }
+
```

Comment:
Under the hood these both use the same data structures: compare `upsertAttrOperation` (line 5 below) with the `write` function in AttributeUpdateOperations.scala - they both construct the same JSON:

```
{
    "op" : "AddUpdateAttribute",
    "attributeName" : "my_attribute",
    "attributeValue" : 42
}
```

So the inside of the first block of `setAttributesOnEntity` builds something the same shape as `getWorkspaceAttributeCalls`, and with a bit of fiddling I think you could get some reuse out of it.

That said: for the sake of getting this PR out the door, I'm willing to come back to this refactor later, as long as we don't make a habit of doing so.